_posts/2021-11-30-pravega-connector-101.md

```diff
@@ -0,0 +1,213 @@
+
+## End-to-end exactly-once semantics
+
+In the early years of big data processing, results from real time stream processing was always considered inaccurate/approximate/speculative. However, this correctness is extremely important in some industries and use cases such as finance. This constraint is mainly coming from the following two problems.
+
+1. The unordered data source in event time 
+2. End-to-end exactly once semantics guarantee
+
+During recent years of development, watermark has been introduced as a tradeoff between correctness and latency, which is now considered as a good solution for Problem 1. 
+
+The second problem is a more difficult one. When we say “exactly-once semantics”, what we mean is that each incoming event affects the final results exactly once. Even in case of a machine or software failure, there’s no duplicate data and no data that goes unprocessed. This is quite hard because of the demanding message acknowledgement and recovery during the fast processing, that's also why some early distributed streaming engines like Storm(without Trident) choose to support "at-least once". Flink is one of the first streaming systems that provided exactly-once semantics internally in an application due to its delicate [checkpoint mechanism](https://www.ververica.com/blog/high-throughput-low-latency-and-exactly-once-stream-processing-with-apache-flink). But to make it end-to-end, it still needs to fix the final stage to apply the semantic to the sink of external message systems that supports commits and rollback. 
+
+Here comes Pravega that supports [transactional writes](https://cncf.pravega.io/docs/latest/transactions/). The idea with a Pravega transaction is that it allows an application to prepare a set of events that can be written "all at once" to a Stream. This allows an application to "commit" a bunch of events atomically. With this Pravega feature making the writes idempotent, it is possible to implement end-to-end exactly-once pipelines together with Flink. 
+
+To build such an end-to-end solution is still not easy, the main difficulty is to have the coordination between Flink and Pravega sink. One common approach for coordinating commits and rollbacks in a distributed system is the two-phase commit protocol. We followed this path, worked together with the Flink community and implemented the sink function in a two-phase commit way coordinated with Flink checkpoints.
+
+Flink community then extracted the common logic of the two-phase commit protocol and provided a general interface `TwoPhaseCommitSinkFunction` (relevant Jira [here](https://issues.apache.org/jira/browse/FLINK-7210)) to make it possible to build end-to-end exactly-once applications with other message system with transaction support including Apache Kafka versions 0.11 and beyond. There is an official Flink [blog](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html) as well to introduce this feature in detail.
```

Comment:
```suggestion
In the early years of big data processing, results from real-time stream processing were always considered inaccurate/approximate/speculative. However, this correctness is extremely important for some use cases and in some industries such as finance. 

This constraint stems mainly from two issues:
- unordered data source in event time 
- end-to-end exactly-once semantics guarantee

During recent years of development, watermarking has been introduced as a tradeoff between correctness and latency, which is now considered a good solution for unordered data sources in event time. 

The guarantee of end-to-end exactly-once semantics is more tricky. When we say “exactly-once semantics”, what we mean is that each incoming event affects the final results exactly once. Even in the event of a machine or software failure, there is no duplicate data and no data that goes unprocessed. This is quite difficult because of the demands of message acknowledgment and recovery during such fast processing and is also why some early distributed streaming engines like Storm(without Trident) chose to support "at-least-once" guarantees. 

Flink is one of the first streaming systems that was able to provide exactly-once semantics due to its delicate [checkpoint mechanism](https://www.ververica.com/blog/high-throughput-low-latency-and-exactly-once-stream-processing-with-apache-flink). But to make it work end-to-end, the final stage needs to apply the semantic to external message system sinks that support commits and rollbacks. 

To work around this problem, Pravega introduced [transactional writes](https://cncf.pravega.io/docs/latest/transactions/). A Pravega transaction allows an application to prepare a set of events that can be written "all at once" to a Stream. This allows an application to "commit" a bunch of events atomically. When writes are idempotent, it is possible to implement end-to-end exactly-once pipelines together with Flink. 

To build such an end-to-end solution requires coordination between Flink and the Pravega sink, which is still challenging. A common approach for coordinating commits and rollbacks in a distributed system is the two-phase commit protocol. We used this protocol and, together with the Flink community, implemented the sink function in a two-phase commit way coordinated with Flink checkpoints.

The Flink community then extracted the common logic from the two-phase commit protocol and provided a general interface `TwoPhaseCommitSinkFunction` ([FLINK-7210](https://issues.apache.org/jira/browse/FLINK-7210)) to make it possible to build end-to-end exactly-once applications with other message systems that have transaction support. This includes Apache Kafka versions 0.11 and above. There is an official Flink [blog post](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html) that describes this feature in detail.
```